wallet/import.go

```diff
@@ -1,13 +1,193 @@
+			return errors.New("invalid master key, must be of the " +
+				"form m/purpose'/coin_type'/account'")
+		}
+		if pubKey.ChildIndex() < hdkeychain.HardenedKeyStart {
+			return errors.New("invalid master key, must be hardened")
+		}
+	} else {
+		if pubKey.Depth() != pubKeyDepth {
+			return errors.New("invalid master key, must be of the " +
+				"form m/purpose'/coin_type'/account'/change/" +
+				"address_index")
+		}
+		if pubKey.ChildIndex() >= hdkeychain.HardenedKeyStart {
+			return errors.New("invalid pulic key, must not be " +
+				"hardened")
+		}
+	}
+
+	return nil
+}
+
+// ImportAccount imports an account backed by a master extended public key. The
+// master key fingerprint denotes the fingerprint of the root key corresponding
+// to the master public key (also known as the key with derivation path m/).
+// This may be required by some hardware wallets for proper identification and
+// signing. The address type can usually be inferred from the key's version, but
+// in the case of legacy versions (xpub, tpub), an address type must be
+// specified as we intend to not support importing BIP-44 keys into the wallet
+// using the legacy pay-to-pubkey (P2PK) scheme.
+func (w *Wallet) ImportAccount(name string, masterPubKey *hdkeychain.ExtendedKey,
+	masterKeyFingerprint uint32, addrType *waddrmgr.AddressType) error {
+
+	// Ensure we have a valid master public key.
+	if err := w.validateExtendedPubKey(masterPubKey, true); err != nil {
+		return err
+	}
+
+	// Determine what key scope the master public key should belong to and
+	// init the account as watch-only.
+	var keyScope waddrmgr.KeyScope
+	pubKeyVersion := pubKeyHDVersion(
+		binary.BigEndian.Uint32(masterPubKey.Version()),
+	)
+	switch pubKeyVersion {
+	case mainnetBIP0044, testnetBIP0044:
+		if addrType == nil {
+			return errors.New("address type must be specified " +
+				"for master public key with legacy version")
+		}
+
+		switch *addrType {
+		case waddrmgr.NestedWitnessPubKey:
+			keyScope = waddrmgr.KeyScopeBIP0049Plus
+		case waddrmgr.WitnessPubKey:
+			keyScope = waddrmgr.KeyScopeBIP0084
+		default:
+			return fmt.Errorf("unsupported address type")
+		}
+
+	case mainnetBIP0049Plus, testnetBIP0049Plus:
+		keyScope = waddrmgr.KeyScopeBIP0049Plus
+
+	case mainnetBIP0084, testnetBIP0084:
+		keyScope = waddrmgr.KeyScopeBIP0084
+
+	default:
+		return fmt.Errorf("unknown version %v", pubKeyVersion)
+	}
+
+	scopedMgr, err := w.Manager.FetchScopedKeyManager(keyScope)
+	if err != nil {
+		return err
+	}
+
+	return walletdb.Update(w.db, func(tx walletdb.ReadWriteTx) error {
+		ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+		_, err = scopedMgr.NewAccountWatchingOnly(
+			ns, name, masterPubKey, masterKeyFingerprint,
+		)
+		return err
+	})
+}
+
+// ImportPubKey imports a single derived public key into the address manager. An
+// optional block can be provided to determine where the rescan should begin.
+func (w *Wallet) ImportPubKey(pubKey *hdkeychain.ExtendedKey) error {
```

Comment:
I'm missing the context on how we're planning on using this, so this is just a side note: If we pass in a default BIP44 extended key (`tpub`), this will cause the address to be of type `PubKeyHash`. So either we only use this with BIP84 extended keys (`vpub`) or need an address type override parameter here as well.

Reply:
I had this fixed but wasn't pushed up at the time of your review. Should be good now, we do the same as in `ImportAccount`.